ext/compressors/zstd/zstd_compress.c

```diff
@@ -0,0 +1,552 @@
+
+	*result_slotsp = slot;
+	*result_lenp = output.pos + sizeof(ZSTD_PREFIX);
+
+	if (0) {
+fail:		*result_slotsp = 0;
+		*result_lenp = 1;
+	}
+	ZSTD_freeCStream(cstream);
+	return (ret);
+}
+
+/*
+ * zstd_pre_size --
+ *	WiredTiger Zstd destination buffer sizing for compression.
+ */
+static int
+zstd_pre_size(WT_COMPRESSOR *compressor, WT_SESSION *session,
+    uint8_t *src, size_t src_len, size_t *result_lenp)
+{
+	(void)compressor;				/* Unused parameters */
+	(void)session;
+	(void)src;
+
+	/*
+	 * Zstd compression runs faster if the destination buffer is sized at
+	 * the upper-bound of the buffer size needed by the compression. Use
+	 * the library calculation of that overhead (plus our overhead).
+	 */
+	*result_lenp = ZSTD_compressBound(src_len) + sizeof(ZSTD_PREFIX);
+	return (0);
+}
+
+/*
+ * zstd_terminate --
+ *	WiredTiger Zstd compression termination.
+ */
+static int
+zstd_terminate(WT_COMPRESSOR *compressor, WT_SESSION *session)
+{
+	(void)session;					/* Unused parameters */
+
+	free(compressor);
+	return (0);
+}
+
+/*
+ * zstd_add_compressor --
+ *	Add a Zstd compressor.
+ */
+static int
+zstd_add_compressor(WT_CONNECTION *connection, int raw, const char *name)
+{
+	ZSTD_COMPRESSOR *zstd_compressor;
+
+	/*
+	 * There are two almost identical Zstd compressors: one using raw
+	 * compression to target a specific block size, and one without.
+	 */
+	if ((zstd_compressor = calloc(1, sizeof(ZSTD_COMPRESSOR))) == NULL)
+		return (errno);
+
+	zstd_compressor->compressor.compress = zstd_compress;
+	zstd_compressor->compressor.compress_raw = raw ?
+	    zstd_compress_raw : NULL;
+	zstd_compressor->compressor.decompress = zstd_decompress;
+	zstd_compressor->compressor.pre_size = zstd_pre_size;
+	zstd_compressor->compressor.terminate = zstd_terminate;
+
+	zstd_compressor->wt_api = connection->get_extension_api(connection);
+
+	/*
+	 * Zstd's sweet-spot is better compression than zlib at significantly
+	 * faster compression/decompression speeds. LZ4 and snappy are faster
+	 * than zstd, but have worse compression ratios. Applications wanting
+	 * faster compression/decompression with worse compression will select
+	 * LZ4 or snappy, so we configure zstd for better compression.
+	 *
+	 * From the zstd github site, default measurements of the compression
+	 * engines we support, listing compression ratios with compression and
+	 * decompression speeds:
+	 *
+	 *	Name	Ratio	C.speed	D.speed
+	 *			MB/s	MB/s
+	 *	zstd	2.877	330	940
+	 *	zlib	2.730	95	360
+	 *	LZ4	2.101	620	3100
+	 *	snappy	2.091	480	1600
+	 *
+	 * Set the zstd compression level to 3: according to the zstd web site,
+	 * that reduces zstd's compression speed to around 200 MB/s, increasing
+	 * the compression ratio to 3.100 (close to zlib's best compression
+	 * ratio). In other words, position zstd as a zlib replacement, having
+	 * similar compression at much higher compression/decompression speeds.
+	 */
+	zstd_compressor->compression_level = 3;
```

Comment:
Maybe (for initial testing at least) we should have "zstd-1", "zstd-3", "zstd-8", "zstd-19", each with a different compression level?